models/balancer/arbitrum/balancer_v2_arbitrum_bpt_prices.sql

```diff
@@ -1,66 +1,160 @@
+        unique_key = ['blockchain', 'hour','contract_address'],
+        post_hook = '{{ expose_spells(\'["arbitrum"]\',
                                     "project",
                                     "balancer_v2",
-                                    \'["victorstefenon"]\') }}'
+                                    \'["victorstefenon", "thetroyharris"]\') }}'
     )
 }}
 
-WITH bpt_trades AS (
-    SELECT
-        block_time,
-        bpt_address,
-        bpt_amount_raw,
-        bpt_amount_raw / POWER(10, COALESCE(erc20a.decimals, 18)) AS bpt_amount,
-        token_amount_raw,
-        token_amount_raw / POWER(10, erc20b.decimals) AS token_amount,
-        p.price * token_amount_raw / POWER(10, erc20b.decimals) AS usd_amount
-    FROM (
-        SELECT
-            t.evt_block_time AS block_time,
-            CASE 
-                WHEN t.tokenIn = SUBSTRING(t.poolId, 0, 42) THEN t.tokenIn
-                ELSE t.tokenOut
-            END AS bpt_address,
-            CASE 
-                WHEN t.tokenIn = SUBSTRING(t.poolId, 0, 42) THEN t.amountIn
-                ELSE t.amountOut
-            END AS bpt_amount_raw,
-            CASE 
-                WHEN t.tokenIn = SUBSTRING(t.poolId, 0, 42) THEN t.tokenOut
-                ELSE t.tokenIn
-            END AS token_address,
+WITH
+    bpt_trades AS (
+        SELECT evt_tx_hash FROM {{ source('balancer_v2_arbitrum', 'Vault_evt_Swap') }} v
+        WHERE CAST(tokenIn AS VARCHAR(66)) = SUBSTRING(CAST(poolId AS VARCHAR(66)), 1, 42)
+        OR CAST(tokenOut AS VARCHAR(66)) = SUBSTRING(CAST(poolId AS VARCHAR(66)), 1, 42) 
+        {% if is_incremental() %}
+        AND v.evt_block_time >= date_trunc("day", now() - interval '1 week')
+        {% endif %} 
+    ), 
+    
+    all_trades_given_bpt_tx AS (
+        SELECT * FROM {{ source('balancer_v2_arbitrum', 'Vault_evt_Swap') }} 
+        WHERE evt_tx_hash IN (TABLE bpt_trades)
+    ), 
+    
+    all_trades_info AS (
+        SELECT 
+            a.evt_tx_hash AS tx_hash,
+            a.evt_block_time AS block_time,
+            a.evt_block_number AS block_number,
+            CAST(a.poolId AS VARCHAR(66)) AS pool_id,
+            SUBSTRING(CAST(a.poolId AS VARCHAR(66)), 1, 42) AS bpt_address,
+            CAST(a.tokenIn AS VARCHAR(66)) AS token_in,
+            CAST(a.amountIn AS DOUBLE) AS amount_in,
+            CAST(a.tokenOut AS VARCHAR(66)) AS token_out,
+            CAST(a.amountOut AS DOUBLE) AS amount_out,
+            p1.price AS token_in_p,
+            COALESCE(p1.symbol, t1.symbol) AS token_in_sym,
+            COALESCE(p1.decimals, t1.decimals) AS token_in_decimals,
+            p2.price AS token_out_p,
+            COALESCE(p2.symbol, t2.symbol) AS token_out_sym,
+            COALESCE(p2.decimals, t2.decimals) AS token_out_decimals
+        FROM all_trades_given_bpt_tx a
+        LEFT JOIN {{ source ('prices', 'usd') }} p1 ON p1.contract_address = a.tokenIn AND p1.blockchain = 'arbitrum' 
```

Comment:
we can add "if is incremental, filter last 7 days" on prices source too -- should help efficiency, as prices is large source

Reply:
and filter can just be in the join conditions itself

Reply:
Ok I'll add this filter too, but isn't it already being filtered by the `AND  p1.minute = date_trunc('minute', a.evt_block_time)` since `evt_block_time` has been filtered for the last 7 days?

Reply:
i think with it being a left join, it just helps reduce the number of rows to look for in that `minute = minute` clause. can't hurt to add 👍